crypto/ec/ec_key.c

```diff
@@ -195,59 +196,87 @@ int ossl_ec_key_gen(EC_KEY *eckey)
     return eckey->group->meth->keygen(eckey);
 }
 
+/*
+ * ECC Key generation.
+ * See SP800-56AR3 5.6.1.2.2 "Key Pair Generation by Testing Candidates"
+ *
+ * Params:
+ *     eckey An EC key object that contains domain params. The generated keypair
+ *           is stored in this object.
+ * Returns 1 if the keypair was generated or 0 otherwise.
+ */
 int ec_key_simple_generate_key(EC_KEY *eckey)
 {
     int ok = 0;
     BN_CTX *ctx = NULL;
-    BIGNUM *priv_key = NULL;
+    BIGNUM *priv_key = NULL, *upper = NULL;
     const BIGNUM *order = NULL;
     EC_POINT *pub_key = NULL;
+    const EC_GROUP *group = eckey->group;
+
+    /*
+     * Step (9): If there is an error return an invalid keypair.
+     * This is done by invalidating any existing key pair, and only
+     * setting the returned keypair into the key if there is no error.
+     */
+    BN_free(eckey->priv_key);
```

Comment:
Does this need to be cleared before freeing?

Reply:
Also, this is changing the existing behaviour of the library not only in `FIPS_MODE`: before `eckey->priv_key` would have been reused if already allocated. What are the performance implications of this change?

> This is done by invalidating any existing key pair, and only setting the returned keypair into the key if there is no error.

Maybe we can design a better strategy to achieve the same behaviour?

Reply:
I used BN_secure_new() to allocate, which I thought does this automagically.

Ok. Good point. But it should at least not leave the old keypair in the key..

Reply:
> I used BN_secure_new() to allocate, which I thought does this automagically.

Yes I noticed this, but I am not sure that every `EC_KEY` object that could be fed to this function has necessarily been populated by this function. Maybe it is, but we should make sure of it.

Reply:
Ok.
That makes me wonder if it should fail if any private key doesnt use BN_secure_new.

Reply:
I would personally wish for such a check, but that could be an API breaking change!

Reply:
you can't do that because on many systems the secure heap needs root privs.

Reply:
As @richsalz noted, secure allocs can be normal ones on some platforms.

A clear is required and mandating secure allocations isn't enforceable.